pkg/storage/consul/consul_helper.go

```diff
@@ -0,0 +1,706 @@
+}
+
+func init() {
+	metrics.Register()
+	gob.Register(watch.Event{})
+	gob.Register(api.Pod{})
+}
+
+func NewConsulStorage(client consulapi.Client, codec runtime.Codec, prefix string, config consulapi.Config) storage.Interface {
+	//ensure we have some prefix otherwise watcher on / will fail because this results in an empty key to watch
+	if prefix == "" {
+		prefix = fmt.Sprintf("pref%d", rand.Intn(800000))
+	}
+	return &consulHelper{
+		Client:             client,
+		Config:             config,
+		codec:              codec,
+		consulKv:           client.KV(),
+		versioner:          etcd.APIObjectVersioner{},
+		pathPrefix:         path.Join("/", prefix),
+		eventHistoryPrefix: "past_events",
+	}
+}
+
+func (h *consulHelper) extractObj(kvPair *consulapi.KVPair, inErr error, objPtr runtime.Object, ignoreNotFound bool, key string) error {
+	if inErr != nil || kvPair == nil || len(kvPair.Value) == 0 {
+		if ignoreNotFound {
+			v, err := conversion.EnforcePtr(objPtr)
+			if err != nil {
+				return err
+			}
+			v.Set(reflect.Zero(v.Type()))
+			return nil
+		} else if inErr != nil {
+			return inErr
+		}
+
+		return toStorageErr(NewNotFoundError(), key, 0)
+	}
+
+	out, gvk, err := h.codec.Decode(kvPair.Value, nil, objPtr)
+	if err != nil {
+		return toStorageErr(err, key, 0)
+	}
+	if out != objPtr {
+		return toStorageErr(NewInvalidObjectError(fmt.Sprintf("unable to decode object %s into %v", gvk.String(), reflect.TypeOf(objPtr))), key, 0)
+	}
+
+	// being unable to set the version does not prevent the object from being extracted
+	_ = h.versioner.UpdateObject(objPtr, kvPair.ModifyIndex)
+	return err
+}
+
+func (h *consulHelper) prefixConsulKey(key string) string {
+	if strings.HasPrefix(key, h.pathPrefix) {
+		return key
+	}
+	return path.Join(h.pathPrefix, key)
+}
+
+func (h *consulHelper) Backends(ctx context.Context) []string {
+	if ctx == nil {
+		glog.Errorf("Context is nil")
+	}
+
+	peers, err := h.Client.Status().Peers()
+	if err != nil {
+		return []string{}
+	}
+
+	return peers
+}
+
+func (h *consulHelper) Versioner() storage.Versioner {
+	return h.versioner
+}
+
+func (h *consulHelper) Create(ctx context.Context, key string, obj, out runtime.Object, ttl uint64) error {
+	trace := util.NewTrace("consulHelper::Create " + getTypeName(obj))
+	defer trace.LogIfLong(250 * time.Millisecond)
+	if ctx == nil {
+		glog.Errorf("Context is nil")
+	}
+
+	originalKey := key
+
+	key = h.prefixConsulKey(key)
+	key = h.transformKeyName(key)
+
+	data, err := runtime.Encode(h.codec, obj)
+	trace.Step("Object encoded")
+	if err != nil {
+		return toStorageErr(NewInvalidObjectError("Encoding object failed: "+err.Error()), key, 0)
+	}
+
+	if version, err := h.versioner.ObjectResourceVersion(obj); err == nil && version != 0 {
+		return toStorageErr(NewInvalidObjectError("resourceVersion may not be set on objects to be created"), key, 0)
+	}
+	trace.Step("Version checked")
+
+	//check if key already exists
+	storedPair, _, err := h.consulKv.Get(key, nil)
```

Comment:
Consul does support CAS. I could depend on the `obj.ObjectMeta.RessourceVersion` for this. However this would bring up an error but maybe the "wrong" one.
Consul does not complain when a key already exists. That's why I added this extra `Get` to check for this case explicitly and return an `NewExistsError` in case